eventsourcing/src/main/java/org/axonframework/eventsourcing/annotation/ConstructorBasedEventSourcedEntityFactory.java

```diff
@@ -17,37 +17,111 @@
 package org.axonframework.eventsourcing.annotation;
 
 import jakarta.annotation.Nonnull;
-import org.axonframework.common.infra.ComponentDescriptor;
-import org.axonframework.common.infra.DescribableComponent;
+import org.axonframework.common.ReflectionUtils;
+import org.axonframework.eventhandling.EventMessage;
 
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Function;
-
-import static org.axonframework.common.ConstructorUtils.factoryForTypeWithOptionalArgument;
+import java.lang.reflect.Constructor;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
 
 /**
  * {@link EventSourcedEntityFactory} implementation which uses a constructor to create a new instance of an entity. The
- * constructor can either have a single argument of the same type as the identifier, or no arguments at all.
+ * constructor can either have a single argument of the same type as the identifier, a single argument of the first
+ * event payload type, or no arguments at all. Note that arguments can not be combined, so a constructor with both an ID
+ * and an event payload is not supported.
  *
+ * @param <E> The type of the entity to create.
  * @author Mitchell Herrijgers
  * @since 5.0.0
  */
-public class ConstructorBasedEventSourcedEntityFactory
-        implements EventSourcedEntityFactory<Object, Object>, DescribableComponent {
+public class ConstructorBasedEventSourcedEntityFactory<E> implements EventSourcedEntityFactory<Object, E> {
```

Comment:
Note: This is, essentially, reflection. A follow-up PR will re-think this method. 
Feel free to review, but it will be replaced by the `ReflectionBasedEntityFactory` in the next PR.